src/content/app-architecture/case-study/ui-layer.md

```diff
@@ -0,0 +1,652 @@
+  
+  final BookingRepository _bookingRepository;
+  final UserRepository _userRepository;
+
+  User? _user;
+  User? get user => _user;
+
+  List<BookingSummary> _bookings = [];
+  UnmodifiableListView<BookingSummary> get bookings => UnmodifiableListView(_bookings);
+
+// ... 
+}
+```
+
+As mentioned, the UI state should be immutable.
+This is a crucial part of bug-free software.
+
+The compass app uses the [freezed][] package to 
+enforce immutability on data classes. For example, 
+the following code shows the `User` class definition. 
+`freezed` provides deep immutability, 
+and generates the code for useful methods like 
+`[copyWith](freezed.copywoith)`and `[toJson](jsonseralizable.toJson)`.
+
+```dart title=user.dart
+@freezed
+class User with _$User {
+  const factory User({
+    /// The user's name.
+    required String name,
+
+    /// The user's picture URL.
+    required String picture,
+  }) = _User;
+
+  factory User.fromJson(Map<String, Object?> json) => _$UserFromJson(json);
+}
+
+```
+
+:::note
+In the ViewModel example, 
+two objects are needed to render the view. 
+As the UI state for any given model grows in complexity, 
+a ViewModel may have many more pieces of data from 
+many more repositories exposed to the view. 
+In some cases, 
+you may want to create objects that specifically represent the UI state. 
+For example, I could create a class called `HomeUiState`.
+:::
+
+
+### Updating UI state
+
+In addition to storing state, 
+ViewModels need to tell Flutter to re-render views when 
+the data layer provides a new state. 
+In the Compass app, ViewModels extend [`ChangeNotifier`][] to achieve this.
+
+```dart title=home_viewmodel.dart
+class HomeViewModel [!extends ChangeNotifier!] {
+  HomeViewModel({
+   required BookingRepository bookingRepository, 
+   required UserRepository userRepository,
+  }) : _bookingRepository = bookingRepository,
+      _userRepository = userRepository;
+  final BookingRepository _bookingRepository;
+  final UserRepository _userRepository;
+
+  User? _user;
+  User? get user => _user;
+
+  List<BookingSummary> _bookings = [];
+  List<BookingSummary> get bookings => _bookings;
+
+  // ... 
+}
+```
+
+`HomeViewModel.user` is a public member that the view depends on. 
+When new data flows from the data layer and 
+new state needs to be emitted, [`notifyListeners`][] is called.
+
+<figure>
+
+![A screenshot of the booking screen of the compass app.](/assets/images/docs/app-architecture/case-study/mvvm-case-study-update-ui-steps.png)
+
+    <figcaption style="font-style: italic">
+This figure shows from a high-level how new data in the repository
+propagates up to the UI layer and triggers a re-build of your Flutter widgets.
+    </figcaption>
+</figure>
+
+1. New state is provided to the ViewModel from a Repository
+2. The ViewModel updates its UI state to reflect the new data
+3. ViewModel.notifyListeners is called, alerting the View of new UI State
+4. The View (widget) re-renders.
```

Comment:
If possible, lists should use [parallel construction](https://developers.google.com/style/lists?hl=en#parallel-syntax) and have consistent ending punctuation.

```suggestion
1. New state is provided to the ViewModel from a Repository.
2. The ViewModel updates its UI state to reflect the new data.
3. ViewModel.notifyListeners is called, alerting the View of new UI State.
4. The View (widget) re-renders.
```